[MM-57080] Add Channel-specific notification sound options in new Notifications settings modal

M-ZubairAhmed opened the pull request:
#### Summary
<!--
A description of what this pull request does, as well as QA test steps (if applicable and if not already added to the Jira ticket).
-->

#### Ticket Link
Jira https://mattermost.atlassian.net/browse/MM-57080


#### Screenshots
![CleanShot 2024-07-22 at 21 48 14](https://github.com/user-attachments/assets/ff087867-0aa0-43b9-a4e1-50335ec7f815)

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
Channel specific message notification sounds configuration added.
```

devinbinnie commented on webapp/channels/src/components/widgets/modals/components/react_select_item.tsx:
```diff
@@ -9,12 +10,17 @@ import type {BaseSettingItemProps} from './base_setting_item';
-    value: number;
-    label: string;
+    value: string;
+    label: string | ReactNode;
```
`ReactNode` includes `string`
```suggestion
    label: ReactNode;
```

M-ZubairAhmed commented:
@abhijit-singh can you please check out this one when you get time

abhijit-singh reviewed (changes requested):
Looks good @M-ZubairAhmed. Noticed a couple of issues in the functionality and one minor cosmetic request:
1. For a new user, if they have never changed the `notification sound` setting in their account settings, then the notification sound setting under channel notification preferences does not have any sound selected. Ideally it should have `Bing` selected by default similar to the global setting. 

 Also, if you change it to some other sound at this point, the `Reset to default` button appears at the top of the section but it doesn't really do anything. 
 
  Note that both of these things work perfectly AFTER I change the global notification sound setting even once. 
  
  https://github.com/user-attachments/assets/58ab9edd-1cff-4286-b15d-a3b405a89516

2. Changing the sound in the channel notification preferences does not play it like it does in the global settings. Ideally, the sound should be played once when selected so that the user knows what it actually sounds like.

3. Can we reduce the width of the input box to ~224px to match with the design? Currently it looks too stretched out and it doesn't have to be.

yasserfaraazkhan reviewed (approved):
Just the 1st and 2nd points shared by Abhijeet.

The Sounds work in channel, GMs

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
> Looks good @M-ZubairAhmed. Noticed a couple of issues in the functionality and one minor cosmetic request:
> 
>     1. For a new user, if they have never changed the `notification sound` setting in their account settings, then the notification sound setting under channel notification preferences does not have any sound selected. Ideally it should have `Bing` selected by default similar to the global setting.
> 
> 
> Also, if you change it to some other sound at this point, the `Reset to default` button appears at the top of the section but it doesn't really do anything.
> 
> Note that both of these things work perfectly AFTER I change the global notification sound setting even once.
> Screen.Recording.2024-07-30.at.11.29.47.AM.mov
> 
>     2. Changing the sound in the channel notification preferences does not play it like it does in the global settings. Ideally, the sound should be played once when selected so that the user knows what it actually sounds like.
> 
>     3. Can we reduce the width of the input box to ~224px to match with the design? Currently it looks too stretched out and it doesn't have to be.

@abhijit-singh all the points mentioned are now fixed

abhijit-singh commented:
Thanks @M-ZubairAhmed. Those seem to be fixed but there's another thing that seems to be happening - 

When I did a fresh sign up of a new user, I could see the `Reset to default` button in Desktop Notifications under Channel Notification Preferences even though I had not changed anything from the default anywhere. Clicking on the button does not seem to do anything and the button does not go away. Ideally, the `Reset to default` button should not be there when the channel settings are matching with the global settings. 


https://github.com/user-attachments/assets/aa5cae4a-a732-4528-82be-b6f073963a9f

M-ZubairAhmed commented:
> Thanks @M-ZubairAhmed. Those seem to be fixed but there's another thing that seems to be happening -
> 
> When I did a fresh sign up of a new user, I could see the `Reset to default` button in Desktop Notifications under Channel Notification Preferences even though I had not changed anything from the default anywhere. Clicking on the button does not seem to do anything and the button does not go away. Ideally, the `Reset to default` button should not be there when the channel settings are matching with the global settings.
> Screen.Recording.2024-08-05.at.4.38.00.PM.mov


I thought i fixed that but i didnt thanks for detailed analysis i fixed it now :)

M-ZubairAhmed commented:
@yasserfaraazkhan adding you again since a lot of code change have happened after you previously approved

yasserfaraazkhan commented:
/e2e-tests

yasserfaraazkhan reviewed (approved):
Thank you @M-ZubairAhmed 
can you please check `channel_notifications_spec.js` e2e test case?

abhijit-singh commented:
Thanks @M-ZubairAhmed. And I'm sorry but I found one more issue, not sure if it was not there before the most recent changes or I just missed it. Here's the issue:
- If I change the notification sound in my global notification setting, it does not get updated on existing channels and neither does it apply to a newly created channel. Hence, I still kept hearing the old notification sound for most messages.

https://github.com/user-attachments/assets/a54ce919-0d1a-4e8a-8a72-e1b24ac2693f

@yasserfaraazkhan Can you confirm if you see this issue?

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/e2e-test

yasserfaraazkhan commented:
/e2e-test

M-ZubairAhmed commented:
@abhijit-singh the issue mentioned above regarding the 

> If I never manually override my channel notification preferences OR when I click on Reset to default and Save, it should then keep my channel setting in sync with the global setting, until I manually change it again. The problem can be seen when the channel setting does not get updated even after changing the global setting.

Is opened into another ticket since this is same not only for channel notification sound but for other channel notification settings as well. Can you re verify in keeping above known bug

M-ZubairAhmed commented:
> E2E test run is starting for commit `d4b20d47e9fd7a6dc0ba653547468a37bddb14d4`. You can check its progress by either:
> 
>     * Looking at the corresponding commit status, which will be available in a few moments: `E2E Tests/test`.
> 
>     * Looking at the [E2E test's Workflow Run](https://github.com/mattermost/mattermost/actions/runs/10390676789), with Run ID `10390676789`.
> 
>     * Looking at the [E2E test's Cycle on the Automation Dashboard](https://automation-dashboard.vercel.app/cycle/10390676789).

E2E shows no failures

M-ZubairAhmed commented:
/update-branch

abhijit-singh reviewed (approved):
Thanks @M-ZubairAhmed! Looks good to me apart from the `Reset to default` issue which we can tackle separately.

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/update-branch

Movion commented:
I can do the QA testing 👍

lindy65 commented:
Assigned to you for testig @Movion - let me know if you have any questions

lindy65 commented:
Hi @Movion - how's the QA review going?  Let me know if you have any questions :)

Movion commented:
> Hi @Movion - how's the QA review going? Let me know if you have any questions :)

Heo @lindy65, I have tested all the options I know of and everything looks good.
As the number of channels and settings increases, it could quickly become confusing, which is why you should consider creating some kind of "overview". But that is not something that is important for the current PR.

The functions as presented work without any problems.

lindy65 reviewed (approved):
Thanks @M-ZubairAhmed - I have also reviewed the PR, taking comments into consideration and, apart from the "Reset to default" issue that is being tracked in a separate issue, everything looks good to me 👍

lindy65 commented:
Awesome @Movion! Thanks for your contribution and testing. Would love your help on future PR QA review requests :)

lindy65 commented:
>As the number of channels and settings increases, it could quickly become confusing, which is why you should consider creating some kind of "overview".

@M-ZubairAhmed - will you create a ticket to consider `@movion's` suggestion for this improvement in a later iteration?

amyblais commented:
/cherry-pick release-9.5

mattermost-build commented:
Cherry pick is scheduled.

mattermost-build commented:
Error trying doing the automated Cherry picking. Please do this manually

```
+++ Updating remotes...
Fetching upstream
hostfile_replace_entries: mkstemp: Read-only file system
update_known_hosts: hostfile_replace_entries failed for /app/.ssh/known_hosts: Read-only file system
Fetching upstream
hostfile_replace_entries: mkstemp: Read-only file system
update_known_hosts: hostfile_replace_entries failed for /app/.ssh/known_hosts: Read-only file system
+++ Updating remotes done...
+++ Creating local branch automated-cherry-pick-of-mattermost-#27730-upstream-release-9.5-1724438627
Switched to a new branch 'automated-cherry-pick-of-mattermost-#27730-upstream-release-9.5-1724438627'
Branch 'automated-cherry-pick-of-mattermost-#27730-upstream-release-9.5-1724438627' set up to track remote branch 'release-9.5' from 'upstream'.

+++ About to attempt cherry pick of PR #27730 with merge commit b0d58bb65ea2436bf0bb1a5390f67352c2a65fe5.

Auto-merging e2e-tests/cypress/tests/integration/channels/collapsed_reply_threads/channel_notifications_spec.js
CONFLICT (content): Merge conflict in e2e-tests/cypress/tests/integration/channels/collapsed_reply_threads/channel_notifications_spec.js
Auto-merging webapp/channels/src/components/channel_notifications_modal/__snapshots__/channel_notifications_modal.test.tsx.snap
Auto-merging webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx
CONFLICT (content): Merge conflict in webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx
Auto-merging webapp/channels/src/components/channel_notifications_modal/utils.tsx
CONFLICT (content): Merge conflict in webapp/channels/src/components/channel_notifications_modal/utils.tsx
CONFLICT (modify/delete): webapp/channels/src/components/user_settings/notifications/desktop_notification_sounds_setting/index.tsx deleted in HEAD and modified in b0d58bb65e ([MM-57080] Add Channel-specific notification sound options in new Notifications settings modal (#27730)).  Version b0d58bb65e ([MM-57080] Add Channel-specific notification sound options in new Notifications settings modal (#27730)) of webapp/channels/src/components/user_settings/notifications/desktop_notification_sounds_setting/index.tsx left in tree.
Auto-merging webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx
CONFLICT (content): Merge conflict in webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx
Auto-merging webapp/channels/src/components/widgets/modals/components/base_setting_item.scss
Auto-merging webapp/channels/src/components/widgets/modals/components/base_setting_item.tsx
CONFLICT (content): Merge conflict in webapp/channels/src/components/widgets/modals/components/base_setting_item.tsx
Auto-merging webapp/channels/src/i18n/en.json
error: could not apply b0d58bb65e... [MM-57080] Add Channel-specific notification sound options in new Notifications settings modal (#27730)
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".

+++ Conflicts detected:

UU e2e-tests/cypress/tests/integration/channels/collapsed_reply_threads/channel_notifications_spec.js
UU webapp/channels/src/components/channel_notifications_modal/channel_notifications_modal.tsx
UU webapp/channels/src/components/channel_notifications_modal/utils.tsx
UU webapp/channels/src/components/user_settings/notifications/user_settings_notifications.tsx
UU webapp/channels/src/components/widgets/modals/components/base_setting_item.tsx
Aborting.

+++ Aborting in-progress git cherry-pick.

+++ Returning you to the master branch and cleaning up.

```

amyblais commented:
@M-ZubairAhmed Please help open a manual cherry-pick PR for ``release-9.5`` branch.

M-ZubairAhmed commented:
@Movion @lindy65 

Can you please explain more about what you mean by 

> As the number of channels and settings increases, it could quickly become confusing, which is why you should consider creating some kind of "overview"

Movion commented:
@M-ZubairAhmed 
It could happen that after a certain time many channels with different sounds are set. Currently there is only the possibility to check the setting directly in the channels. My suggestion is to create an overview page of the channels and their settings to counteract the complexity. 
Whether this makes sense or not is up to the product owners. It's just an idea that came to me while I was testing.

lindy65 commented:
@M-ZubairAhmed - I agree with `@movion` that having a modal that lists notification sounds for all channels as well as the default notification sound that's set in Settings > Notifications > Desktop notification sounds could be a nice improvement in a future iteration.  Perhaps also being possible to change the notification sound for each channel from that modal.

M-ZubairAhmed commented:
@Movion would you be open for creating a github issue with your idea suggestion?
